File: stage0/src/lexer.py
"""Fuse Stage 0 — Lexer.

Converts source text to a flat token stream.
Tracks line and column for every token.
"""

from __future__ import annotations
from fuse_token import Token, TokenType, KEYWORDS
from errors import LexerError


class Lexer:
    def __init__(self, source: str, filename: str = "<stdin>"):
        self.source = source
        self.filename = filename
        self.pos = 0
        self.line = 1
        self.col = 1
        self.tokens: list[Token] = []

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def tokenize(self) -> list[Token]:
        while self.pos < len(self.source):
            self._skip_whitespace_and_comments()
            if self.pos >= len(self.source):
                break

            ch = self._current()

            if ch.isdigit():
                self._read_number()
            elif ch == '"':
                self._read_string()
            elif ch == "f" and self._peek(1) == '"':
                self._read_fstring()
            elif ch.isalpha() or ch == "_":
                self._read_identifier()
            else:
                self._read_operator()

        self.tokens.append(Token(TokenType.EOF, None, self.line, self.col))
        return self.tokens

    # ------------------------------------------------------------------
    # Character helpers
    # ------------------------------------------------------------------

    def _current(self) -> str:
        return self.source[self.pos] if self.pos < len(self.source) else "\0"

    def _peek(self, offset: int = 1) -> str:
        p = self.pos + offset
        return self.source[p] if p < len(self.source) else "\0"

    def _advance(self) -> str:
        ch = self.source[self.pos]
        self.pos += 1
        if ch == "\n":
            self.line += 1
            self.col = 1
        else:
            self.col += 1
        return ch

    def _error(self, msg: str) -> LexerError:
        return LexerError(msg, self.filename, self.line, self.col)

    # ------------------------------------------------------------------
    # Whitespace & comments
    # ------------------------------------------------------------------

    def _skip_whitespace_and_comments(self):
        while self.pos < len(self.source):
            ch = self._current()
            if ch in " \t\r\n":
                self._advance()
            elif ch == "/" and self._peek() == "/":
                while self.pos < len(self.source) and self._current() != "\n":
                    self._advance()
            else:
                break

    # ------------------------------------------------------------------
    # Numbers
    # ------------------------------------------------------------------

    def _read_number(self):
        start_line, start_col = self.line, self.col
        num = ""
        is_float = False

        while self.pos < len(self.source):
            ch = self._current()
            if ch.isdigit():
                num += self._advance()
            elif ch == ".":
                nxt = self._peek()
                if nxt == ".":          # .. range operator
                    break
                if nxt.isalpha() or nxt == "_":  # method call on int
                    break
                is_float = True
                num += self._advance()
            else:
                break

        if is_float:
            self.tokens.append(Token(TokenType.FLOAT, float(num), start_line, start_col))
        else:
            self.tokens.append(Token(TokenType.INT, int(num), start_line, start_col))

    # ------------------------------------------------------------------
    # Strings
    # ------------------------------------------------------------------

    def _read_string(self):
        start_line, start_col = self.line, self.col
        self._advance()  # skip opening "
        result = self._read_string_body()
        self.tokens.append(Token(TokenType.STRING, result, start_line, start_col))

    def _read_string_body(self) -> str:
        start_line, start_col = self.line, self.col
        result = ""
        while self.pos < len(self.source) and self._current() != '"':
            if self._current() == "\\":
                self._advance()
                result += self._read_escape()
            else:
                result += self._advance()
        if self.pos >= len(self.source):
            raise self._error("Unterminated string")
        self._advance()  # skip closing "
        return result

    def _read_escape(self) -> str:
        if self.pos >= len(self.source):
            raise self._error("Unterminated escape sequence")
        ch = self._advance()
        return {"n": "\n", "t": "\t", "\\": "\\", '"': '"',
                "{": "{", "}": "}"}.get(ch, "\\" + ch)

    # ------------------------------------------------------------------
    # F-strings
    # ------------------------------------------------------------------

    def _read_fstring(self):
        start_line, start_col = self.line, self.col
        self._advance()  # skip 'f'
        self._advance()  # skip '"'

        parts: list[tuple[str, str]] = []
        current = ""

        while self.pos < len(self.source) and self._current() != '"':
            if self._current() == "{":
                if current:
                    parts.append(("str", current))
                    current = ""
                self._advance()  # skip '{'
                expr = self._read_fstring_expr()
                parts.append(("expr", expr))
            elif self._current() == "\\":
                self._advance()
                current += self._read_escape()
            else:
                current += self._advance()

        if current:
            parts.append(("str", current))

        if self.pos >= len(self.source):
            raise self._error("Unterminated f-string")
        self._advance()  # skip closing "

        self.tokens.append(Token(TokenType.FSTRING, parts, start_line, start_col))

    def _read_fstring_expr(self) -> str:
        """Read the expression inside {...} of an f-string."""
        expr = ""
        depth = 1
        while self.pos < len(self.source) and depth > 0:
            ch = self._current()
            if ch == "{":
                depth += 1
            elif ch == "}":
                depth -= 1
            if depth > 0:
                expr += self._advance()
            else:
                self._advance()  # skip closing '}'
        if depth > 0:
            raise self._error("Unterminated f-string interpolation")
        return expr

    # ------------------------------------------------------------------
    # Identifiers & keywords
    # ------------------------------------------------------------------

    def _read_identifier(self):
        start_line, start_col = self.line, self.col
        ident = ""
        while self.pos < len(self.source) and (self._current().isalnum() or self._current() == "_"):
            ident += self._advance()
        token_type = KEYWORDS.get(ident, TokenType.IDENT)
        self.tokens.append(Token(token_type, ident, start_line, start_col))

    # ------------------------------------------------------------------
    # Operators & delimiters
    # ------------------------------------------------------------------

    _TWO_CHAR: dict[str, TokenType] = {
        "=>": TokenType.FAT_ARROW,
        "->": TokenType.ARROW,
        "?.": TokenType.QUESTION_DOT,
        "?:": TokenType.ELVIS,
        "==": TokenType.EQEQ,
        "!=": TokenType.BANGEQ,
        "<=": TokenType.LTEQ,
        ">=": TokenType.GTEQ,
        "::": TokenType.COLONCOLON,
        "..": TokenType.DOTDOT,
    }

    _ONE_CHAR: dict[str, TokenType] = {
        "?": TokenType.QUESTION,
        "@": TokenType.AT,
        ".": TokenType.DOT,
        ":": TokenType.COLON,
        "=": TokenType.EQUALS,
        "<": TokenType.LT,
        ">": TokenType.GT,
        "+": TokenType.PLUS,
        "-": TokenType.MINUS,
        "*": TokenType.STAR,
        "/": TokenType.SLASH,
        "%": TokenType.PERCENT,
        "|": TokenType.PIPE,
        "(": TokenType.LPAREN,
        ")": TokenType.RPAREN,
        "{": TokenType.LBRACE,
        "}": TokenType.RBRACE,
        "[": TokenType.LBRACKET,
        "]": TokenType.RBRACKET,
        ",": TokenType.COMMA,
        ";": TokenType.SEMICOLON,
    }

    def _read_operator(self):
        start_line, start_col = self.line, self.col
        two = self._current() + self._peek()

        if two in self._TWO_CHAR:
            self._advance()
            self._advance()
            self.tokens.append(Token(self._TWO_CHAR[two], two, start_line, start_col))
            return

        ch = self._current()
        if ch in self._ONE_CHAR:
            self._advance()
            self.tokens.append(Token(self._ONE_CHAR[ch], ch, start_line, start_col))
            return

        raise self._error(f"Unexpected character: {ch!r}")


# ------------------------------------------------------------------
# Standalone entry point
# ------------------------------------------------------------------

if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage: python lexer.py <file.fuse>")
        sys.exit(1)
    with open(sys.argv[1]) as f:
        source = f.read()
    lexer = Lexer(source, sys.argv[1])
    for tok in lexer.tokenize():
        print(tok)

File: stage1/fusec/src/lexer/lexer.rs
use crate::error::FuseError;
use super::token::*;

pub struct Lexer {
    src: Vec<char>,
    pos: usize,
    line: usize,
    col: usize,
    file: String,
}

impl Lexer {
    pub fn new(source: &str, file: &str) -> Self {
        Self { src: source.chars().collect(), pos: 0, line: 1, col: 1, file: file.into() }
    }

    pub fn tokenize(&mut self) -> Result<Vec<Token>, FuseError> {
        let mut tokens = Vec::new();
        loop {
            self.skip_ws();
            if self.pos >= self.src.len() { break; }
            let ch = self.cur();
            if ch.is_ascii_digit() {
                tokens.push(self.number()?);
            } else if ch == '"' {
                tokens.push(self.string()?);
            } else if ch == 'f' && self.peek(1) == Some('"') {
                tokens.push(self.fstring()?);
            } else if ch.is_ascii_alphabetic() || ch == '_' {
                tokens.push(self.ident());
            } else {
                tokens.push(self.operator()?);
            }
        }
        tokens.push(Token { ty: Tok::Eof, line: self.line, col: self.col });
        Ok(tokens)
    }

    // ── helpers ──────────────────────────────────────────────────────
    fn cur(&self) -> char { self.src[self.pos] }
    fn peek(&self, off: usize) -> Option<char> { self.src.get(self.pos + off).copied() }

    fn advance(&mut self) -> char {
        let ch = self.src[self.pos];
        self.pos += 1;
        if ch == '\n' { self.line += 1; self.col = 1; } else { self.col += 1; }
        ch
    }

    fn err(&self, msg: impl Into<String>) -> FuseError {
        FuseError::new(msg, &self.file, self.line, self.col)
    }

    // ── whitespace / comments ────────────────────────────────────────
    fn skip_ws(&mut self) {
        while self.pos < self.src.len() {
            let ch = self.cur();
            if ch == ' ' || ch == '\t' || ch == '\r' || ch == '\n' {
                self.advance();
            } else if ch == '/' && self.peek(1) == Some('/') {
                while self.pos < self.src.len() && self.cur() != '\n' { self.advance(); }
            } else {
                break;
            }
        }
    }

    // ── numbers ──────────────────────────────────────────────────────
    fn number(&mut self) -> Result<Token, FuseError> {
        let (sl, sc) = (self.line, self.col);
        let mut s = String::new();
        let mut is_float = false;
        while self.pos < self.src.len() {
            let ch = self.cur();
            if ch.is_ascii_digit() {
                s.push(self.advance());
            } else if ch == '.' {
                if self.peek(1) == Some('.') { break; }
                if self.peek(1).map_or(false, |c| c.is_ascii_alphabetic() || c == '_') { break; }
                is_float = true;
                s.push(self.advance());
            } else {
                break;
            }
        }
        let ty = if is_float {
            Tok::Float(s.parse::<f64>().map_err(|_| self.err("invalid float"))?)
        } else {
            Tok::Int(s.parse::<i64>().map_err(|_| self.err("invalid int"))?)
        };
        Ok(Token { ty, line: sl, col: sc })
    }

    // ── strings ──────────────────────────────────────────────────────
    fn string(&mut self) -> Result<Token, FuseError> {
        let (sl, sc) = (self.line, self.col);
        self.advance(); // skip "
        let s = self.string_body()?;
        Ok(Token { ty: Tok::Str(s), line: sl, col: sc })
    }

    fn string_body(&mut self) -> Result<String, FuseError> {
        let mut s = String::new();
        while self.pos < self.src.len() && self.cur() != '"' {
            if self.cur() == '\\' {
                self.advance();
                s.push(self.escape()?);
            } else {
                s.push(self.advance());
            }
        }
        if self.pos >= self.src.len() { return Err(self.err("unterminated string")); }
        self.advance(); // skip closing "
        Ok(s)
    }

    fn escape(&mut self) -> Result<char, FuseError> {
        if self.pos >= self.src.len() { return Err(self.err("unterminated escape")); }
        let ch = self.advance();
        Ok(match ch {
            'n' => '\n', 't' => '\t', '\\' => '\\', '"' => '"',
            '{' => '{', '}' => '}',
            _ => ch,
        })
    }

    // ── f-strings ────────────────────────────────────────────────────
    fn fstring(&mut self) -> Result<Token, FuseError> {
        let (sl, sc) = (self.line, self.col);
        self.advance(); // f
        self.advance(); // "
        let mut parts = Vec::new();
        let mut buf = String::new();
        while self.pos < self.src.len() && self.cur() != '"' {
            if self.cur() == '{' {
                if !buf.is_empty() { parts.push(FStringPart::Str(std::mem::take(&mut buf))); }
                self.advance();
                let expr = self.fstring_expr()?;
                parts.push(FStringPart::Expr(expr));
            } else if self.cur() == '\\' {
                self.advance();
                buf.push(self.escape()?);
            } else {
                buf.push(self.advance());
            }
        }
        if !buf.is_empty() { parts.push(FStringPart::Str(buf)); }
        if self.pos >= self.src.len() { return Err(self.err("unterminated f-string")); }
        self.advance(); // "
        Ok(Token { ty: Tok::FString(parts), line: sl, col: sc })
    }

    fn fstring_expr(&mut self) -> Result<String, FuseError> {
        let mut s = String::new();
        let mut depth = 1u32;
        while self.pos < self.src.len() && depth > 0 {
            let ch = self.cur();
            if ch == '{' { depth += 1; }
            if ch == '}' { depth -= 1; }
            if depth > 0 { s.push(self.advance()); } else { self.advance(); }
        }
        if depth > 0 { return Err(self.err("unterminated f-string interpolation")); }
        Ok(s)
    }

    // ── identifiers / keywords ───────────────────────────────────────
    fn ident(&mut self) -> Token {
        let (sl, sc) = (self.line, self.col);
        let mut s = String::new();
        while self.pos < self.src.len() && (self.cur().is_ascii_alphanumeric() || self.cur() == '_') {
            s.push(self.advance());
        }
        let ty = keyword(&s).unwrap_or_else(|| Tok::Ident(s));
        Token { ty, line: sl, col: sc }
    }

    // ── operators / delimiters ───────────────────────────────────────
    fn operator(&mut self) -> Result<Token, FuseError> {
        let (sl, sc) = (self.line, self.col);
        let c1 = self.cur();
        let c2 = self.peek(1);
        let two: String = if let Some(c) = c2 { format!("{c1}{c}") } else { String::new() };

        let two_ty = match two.as_str() {
            "=>" => Some(Tok::FatArrow), "->" => Some(Tok::Arrow),
            "?." => Some(Tok::QuestionDot), "?:" => Some(Tok::Elvis),
            "==" => Some(Tok::EqEq), "!=" => Some(Tok::BangEq),
            "<=" => Some(Tok::LtEq), ">=" => Some(Tok::GtEq),
            "::" => Some(Tok::ColonColon), ".." => Some(Tok::DotDot),
            _ => None,
        };
        if let Some(ty) = two_ty {
            self.advance(); self.advance();
            return Ok(Token { ty, line: sl, col: sc });
        }

        let one_ty = match c1 {
            '?' => Tok::Question, '@' => Tok::At, '.' => Tok::Dot,
            ':' => Tok::Colon, '=' => Tok::Eq, '<' => Tok::Lt, '>' => Tok::Gt,
            '+' => Tok::Plus, '-' => Tok::Minus, '*' => Tok::Star,
            '/' => Tok::Slash, '%' => Tok::Percent, '|' => Tok::Pipe,
            '(' => Tok::LParen, ')' => Tok::RParen,
            '{' => Tok::LBrace, '}' => Tok::RBrace,
            '[' => Tok::LBracket, ']' => Tok::RBracket,
            ',' => Tok::Comma, ';' => Tok::Semicolon,
            _ => return Err(self.err(format!("unexpected character: {c1:?}"))),
        };
        self.advance();
        Ok(Token { ty: one_ty, line: sl, col: sc })
    }
}
